### PR TITLE
Fix bug for undefined-compiler in older webpack configs 

### PR DESCRIPTION
**Pull request by mrnocreativity**
I was having an issue where the plugin would not run and actually break the webpack build. As it turned out, a missing argument was the cause. I added it and rewrote some of the code since it seemed hard to grasp the concepts (for me peronally). Feel free to ignore the rewrite if you don't agree :)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,38 +6,41 @@ class CleanTerminalPlugin {
 
     this.message = message;
     this.onlyInWatchMode = onlyInWatchMode;
+    this.doClearConsole = false;
```
Comment (danillouz) [inline index.js]: What's the purpose of `doClearConsole`?

File: index.js [hunk 2]
```diff
@@ -6,38 +6,41 @@ class CleanTerminalPlugin {
...
-      return compiler.watchMode;
+      return !!compiler.watchMode;
     }
 
-    return (
-      process.env.NODE_ENV !== 'production' &&
-      process.env.options.mode !== 'production'
-    );
+    const isNodeEnvProduction = process.env.NODE_ENV == 'production';
+    const isOptionsModeProduction = process.env.mode && process.env.options.mode == 'production';
```
Comment (danillouz) [inline index.js]: I'm not familiar with `process.env.mode`, but is `process.env.options.mode` only available when `process.env.mode` is `true`?

File: index.js (hunk 1 above)
Comment (mrnocreativity) [reply index.js]: Ow, that needs to go! Will amend the pull request tomorrow to fix this!

File: index.js (hunk 2 above)
Comment (mrnocreativity) [reply index.js]: Quite honestly, I never put any though into the two checks. I'm not sure about where they come from. I suspect the env values could come from flags while the options.mode is what's in the config file.

File: index.js (hunk 1 above)
Comment (mrnocreativity) [reply index.js]: Variable removed in latest commit

**Comment by mrnocreativity** [comment]
Applied linting and removed the dead variable. Let me know if you want me to fix something else as well :)

File: index.js (hunk 2 above)
Comment (danillouz) [reply index.js]: Ah right, I can't find any documentation on if, so I guess we should leave it in then. What do you think?

Comment (mrnocreativity) [reply index.js]: If I'm not mistaken @andykais added that part. Maybe he can explain the internals here?

Comment (andykais) [reply index.js]: I did. Its another mistake of mine I am afraid. That line needs to be changed from `process.env.options.mode` to `compiler.options.mode`. I cannot find documentation on it anywhere, but `compiler.options` is essentially the normalized webpack.config.js object (e.g. its defaults are filled in). So `compiler.options.mode` is how you access the `webpack --mode=` flag, or one defined in the config.

Here is it being used in the [copy-webpack-plugin](https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/src/index.js#L27)

Comment (mrnocreativity) [reply index.js]: Changed as per your feedback. We're now checking if production is either set in the env flags or in the compiler options mode. Running like a kitten :)

Comment (danillouz) [reply index.js]: Ok great, looks good!
Thanks for all the help and input. Nice work!

Going to merge and publish. Let me know if there's anything more to be done, cheers.

**Pull request merged**